service/internal/s3shared/update_endpoint.go

```diff
@@ -0,0 +1,48 @@
+package s3shared
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/awslabs/smithy-go/middleware"
+	"github.com/awslabs/smithy-go/transport/http"
+)
+
+type EnableDualstackMiddleware struct {
+	UseDualstack bool
+	ServiceID    string
+}
+
+// ID returns the middleware ID.
+func (*EnableDualstackMiddleware) ID() string { return "EnableDualstackMiddleware" }
+
+func (u *EnableDualstackMiddleware) HandleSerialize(
+	ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler,
+) (
+	out middleware.SerializeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*http.Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown request type %T", req)
```

Comment:
We use this error so often, maybe it would be helpful to have a defined type in smithy-go, or internally to the SDK even. Would help with consistent error messages probably.

created https://github.com/aws/aws-sdk-go-v2/issues/842 for this